rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,134 @@
+---
+authors: Michael Wilson (michael.wilson@goteleport.com)
+state: draft
+---
+
+# RFD 124 - Plugin static credentials
+
+### Required Approvers
+
+* Engineering @r0mant, @justinas
+* Security @reed
+* Product: @klizhentas
+
+## What
+
+Allow storage of plugin static credentials within Teleport in a safe and secure way.
+
+## Why
+
+As we grow our plugin support, we have discovered that the various services we
+hope to integrate do not offer a uniform mechanism for configuration. For instance,
+Slack uses oauth2, but Okta requires an API token, and Jamf requires a username and
+password.
+
+Today, the plugin houses a credentials object that allows for this sort of storage.
+However, unlike oauth tokens, static credentials pose a much greater risk of exposure,
+and rotating credentials can be a time consuming process.  To reduce the risks associated
+with storing static credentials within Teleport and to allow for easier credentials rotation,
+we would like to create a new resource within Teleport that is handled in an explicitly secure
+way.
+
+## Details
+
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+
+It should be noted that this object should *only* be readable by the auth service to prevent
+exfiltration of these credentials.
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: plugin-name
+spec:
+  lastRotated: "2023-05-09T16:50:52.497874Z"
+  credentials:
+    # Only one of these credential types must be defined at a time.
+    api_token: example-token
+    basic_auth:
+      username: example-user
+      password: example-password
+```
+
+If a plugin requires static credentials, there is a 1-to-1 mapping between `Plugin`
+objects and `PluginStaticCredentials` objects. If a `Plugin` is using oauth, however,
```

Comment:
From what I remember from our discussion yesterday, we may need to store OAuth clientId and clientSecret as well such as in case of Mattermost plugin.

Reply:
I've added in oauth client secret as a static credential.